### PR TITLE
Working with java synchronization

### LINKED ISSUE #5936: Working with java synchronization
#### Proposed title of article
[Languages] Working with java synchronization

#### Proposed article introduction
The synchronization mechanism in java prohibits multiple threads from accessing shared resources. Shared resources include external files, class variables, and database information.
Synchronization is prevalent in multithreaded programming. Synchronized is the keyword that allows your code to run on a single thread at a time without interruption.

#### Key takeaways 
After reading this article, the reader should be able to understand:
-  What is synchronization in java and why do we need it.
- How synchronization works
- How to declare a method as synchronized
- Synchronized method vs Synchronized block 

#### Article quality
This will be an in-depth article covering all the aspects of this concept. It will explain key concepts to readers which is unavailable in the documentations and other platforms. I aim at making it more beginner-friendly simple and easy to understand. More on the coverage will be on the types of synchronization as well as multi-threading with and without synchronization

#### References
N/A

### PR DESCRIPTION
**Pull request by gracy20**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

#### Topic suggestion this closes
This closes #5936

### CODE REVIEW

**Comment by gracy20** [comment]
Hello, @john-amiscaray I have created another PR and done the requested changes. Please check thankyou.

File: content/articles/working-with-java-synchronization/index.md [hunk 1]
```diff
@@ -87,9 +87,9 @@ The output of the thread is : Thread-4 - 5
...
 
-It is evident that the crucial element (code that changes shared resources) must be limited. Additionally, with Java's `synchronized` keyword we can synchronize access to the shared resource. This prevents thread interference during atomic operations. The term "atomic operation" refers to a set of operations that are always performed in unison. All of the atomic actions must be completed at the same time, or none of them can be completed at all.
+It is evident that the crucial element (code that changes shared resources) must be limited. Additionally, with Java's `synchronized keyword` we can synchronize access to the shared resource. This prevents thread interference during atomic operations. The term `atomic operation` refers to a set of operations that are always performed in unison. All of the atomic actions must be completed at the same time, or none of them can be completed at all.
```
Comment (john-amiscaray) [inline content/articles/working-with-java-synchronization/index.md:92]: Backticks (\`) are meant to be used when emphasizing that something is a part of code. For example, you could use it to denote the name of a class, a keyword, the name of a method, etc.  So instead, I'm going to change it to wrap *atomic operation* in asterisks (\*) and change: 

`synchronized keyword` 

to:

 `synchronized` keyword.

File: content/articles/working-with-java-synchronization/index.md [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+public class TestSynchronizedBlock1{
+    public static void main(String[] args){
+        Table obj = new Table();//only one object
+        MyThread1 t1=new MyThread1(obj);
+        MyThread2 t2=new MyThread2(obj);
+        t1.start();
+        t2.start();
+    }
+}
```
Comment (john-amiscaray) [inline content/articles/working-with-java-synchronization/index.md:204]: There still seems to be a race condition here. Most of the time I get the correct output but occasionally I get this:
Capture

File: content/articles/working-with-java-synchronization/index.md [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+
+public class SynchImp {
+	public static void main(String[] args) {
+		Synchronization b1 = new Synchronization();
+		OurThreadExample mt1 = new OurThreadExample(b1, "SECTION");
+		OurThreadExample mt2 = new OurThreadExample(b1, "ENGINEERING");
+		mt1.start();
+		mt2.start();
+	}
+}
```
Comment (john-amiscaray) [inline content/articles/working-with-java-synchronization/index.md:279]: Here too there is a race condition. Occasionally, I get this output:
Capture

**Review by john-amiscaray — CHANGES_REQUESTED** [review]
Your code still has race conditions (see the code below). I believe the cause for it is that while you made it so that only one thread can access the critical section at a time (using the synchronized block/method), the scheduling order of the threads is still unpredictable. Thus, you need to find a way to fix the order of thread execution. I would have a look at the `ScheduledThreadPoolExecutor` class which has methods for you to schedule the timing of classes implementing the `Runnable` interface (which the `Thread` class is).

**Comment by hectorkambow** [comment]
@john-amiscaray Im seeing this is a duplicate - therefore if any more (major) edits are needed we will be closing this PR.

**Comment by john-amiscaray** [comment]
> @john-amiscaray I'm seeing this is a duplicate - therefore if any more (major) edits are needed we will be closing this PR.

@hectorkambow Got it. Though I believe after the fixes I suggested above it should be good. I'm going through it with the author I don't believe it should be that difficult of a fix.

**Comment by gracy20** [comment]
Hello @john-amiscaray @hectorkambow  I'm yet to finish my exams and seems like I'll take a little bit more time than I requested. Therefore I'm considering closing it as per now and reopen it once I'm done.

**Comment by gracy20** [comment]
Hello @john-amiscaray would you please have a look.

**Comment by hectorkambow** [comment]
@gracy20 
FYI
Just going through your PR & topic form. 

We can proceed to get it (reviewed &) published, but it will not be eligible for a payout. (due to the fact that it was closed and re-opened).

Then the fact that the article was re-opened after our transition announcement - that is also a main reason.

The topic form was closed on [May 17th](https://github.com/section-engineering-education/engineering-education/issues/5936#event-6627803039)

& I just noticed that you had closed it May 4th. 

Thank you - and we are working on getting all the articles reviewed and published asap 🚀

**Comment by john-amiscaray** [comment]
I'll try to have a look at this today or tomorrow to get this ready for a final review

**Review by john-amiscaray — APPROVED** [review]
LGTM

**Comment by hectorkambow** [comment]
jumping on this one

**Comment by hectorkambow** [comment]
@gracy20 
👏 Nicely done.

You'll be able to see your live published article on the site (in about 30 mins).

To give your article a boost in page engagement - share it across your social site (LinkedIn/Twitter) and tag the EngEd program so we can also share it and give it some traction as well!
(we will try and do the same once we feature it).

If you have not done so already - consider adding a skill section to your author page. This will go a long way for potential future employers.

**Pull request merged**